File: src/common/FileUtils.cpp
#include "FileUtils.h"

#include <fstream>
#include <iostream>
#include <algorithm>

using namespace std;

vector<Chunk> SplitFileIntoChunks(const string &filepath)
{
    vector<Chunk> chunks;

    // open file in binary mode
    ifstream file(filepath, ios::binary);

    if (!file.is_open())
    {
        cerr << "Error, file cannot be opened " << filepath << endl;
        return chunks;
    }

    int index = 0;

    while (true)
    {
        // create buffer of 1mb
        vector<char> buffer(CHUNK_SIZE);

        file.read(buffer.data(), CHUNK_SIZE);

        size_t bytes_read = file.gcount();

        // if we read something, make chunks
        if (bytes_read > 0)
        {
            Chunk chunk;
            chunk.index = index++;
            chunk.size = bytes_read;
            chunk.data = vector<char>(buffer.begin(), buffer.begin() + bytes_read);
            // todo: fill chunk.hash later using hashing utiliy
            chunks.push_back(chunk);
        }
        if (file.eof())
            break;
    }
    file.close();
    return chunks;
}

bool ReconstructFile(const vector<Chunk> &chunks, const string &output_path)
{
    if (chunks.empty())
    {
        cerr << "Empty chunks. cant reconstruct" << endl;
        return false;
    }

    vector<Chunk> sorted_chunks = chunks;
    sort(sorted_chunks.begin(), sorted_chunks.end(),
         [](const Chunk &a, const Chunk &b)
         {
             return a.index < b.index;
         });

    // check for misisng chunks.
    for (size_t i = 0; i < sorted_chunks.size(); i++)
    {
        if (sorted_chunks[i].index != (int)i)
        {
            cerr << "Error: missing chunk " << i << endl;
            return false;
        }
    }

    // open file from output path in write mode
    ofstream file(output_path, ios::binary);
    if (!file.is_open())
    {
        cerr << "Erorr: file could not be created " << output_path << endl;
        return false;
    }

    // write each chunks data
    for (const Chunk &chunk : sorted_chunks)
    {
        file.write(chunk.data.data(), chunk.size);
    }

    file.close();
    return true;
}

File: src/main/java/com/distributed/storage/common/FileUtils.java
package com.distributed.storage.common;

import java.io.*;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class FileUtils {

    public static final int CHUNK_SIZE = 1048576; // 1MB

    public static List<Chunk> splitFileIntoChunks(String filepath) {
        List<Chunk> chunks = new ArrayList<>();
        File file = new File(filepath);

        if (!file.exists()) {
            System.err.println("Error, file cannot be opened " + filepath);
            return chunks;
        }

        try (FileInputStream fis = new FileInputStream(file)) {
            byte[] buffer = new byte[CHUNK_SIZE];
            int bytesRead;
            int index = 0;

            while ((bytesRead = fis.read(buffer)) != -1) {
                if (bytesRead > 0) {
                    Chunk chunk = new Chunk();
                    chunk.index = index++;
                    chunk.size = bytesRead;
                    // Create exact sized array
                    chunk.data = new byte[bytesRead];
                    System.arraycopy(buffer, 0, chunk.data, 0, bytesRead);
                    chunks.add(chunk);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return chunks;
    }

    public static boolean reconstructFile(List<Chunk> chunks, String outputPath) {
        if (chunks == null || chunks.isEmpty()) {
            System.err.println("Empty chunks. cant reconstruct");
            return false;
        }

        List<Chunk> sortedChunks = new ArrayList<>(chunks);
        sortedChunks.sort(Comparator.comparingInt(a -> a.index));

        // check for missing chunks
        for (int i = 0; i < sortedChunks.size(); i++) {
            if (sortedChunks.get(i).index != i) {
                System.err.println("Error: missing chunk " + i);
                return false;
            }
        }

        try (FileOutputStream fos = new FileOutputStream(outputPath)) {
            for (Chunk chunk : sortedChunks) {
                fos.write(chunk.data);
            }
        } catch (IOException e) {
            System.err.println("Error: file could not be created " + outputPath);
            e.printStackTrace();
            return false;
        }

        return true;
    }
}
